pkg/querier/queryrange/split_by_interval.go

```diff
@@ -253,11 +253,26 @@ func splitByTime(req queryrangebase.Request, interval time.Duration) []queryrang
 	return reqs
 }
 
-func forInterval(interval time.Duration, start, end time.Time, callback func(start, end time.Time)) {
+func forInterval(interval time.Duration, start, end time.Time, endTimeInclusive bool, callback func(start, end time.Time)) {
+	// align the start time by split interval for better query performance of metadata queries and
+	// better cache-ability of query types that are cached.
+	ogStart := start
+	startNs := start.UnixNano()
+	start = time.Unix(0, startNs-startNs%interval.Nanoseconds())
+	firstInterval := true
+
 	for start := start; start.Before(end); start = start.Add(interval) {
 		newEnd := start.Add(interval)
-		if newEnd.After(end) {
+		if !newEnd.Before(end) {
 			newEnd = end
+		} else if endTimeInclusive {
+			newEnd = newEnd.Add(-time.Millisecond)
+		}
```

Comment:
I don't quite understand this.

Reply:
When the end time is inclusive, we do not want the current query's end time to be the same as the start time of the next query. The only queries that have both start and end time inclusive are metadata queries, and without this change, we would still end up querying the same data as the next query.
I will add a comment to make it clear.